File: backend/app/models/deployment.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, DateTime, Text, Integer, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship

from app.core.database import Base


class DeploymentStatus(str, enum.Enum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    CANCELLED = "cancelled"
    SUCCESS = "success"
    FAILED = "failed"
    ROLLING_BACK = "rolling_back"
    ROLLED_BACK = "rolled_back"


class DeploymentStep(str, enum.Enum):
    CLONE = "clone"
    BUILD = "build"
    PUSH = "push"
    DEPLOY = "deploy"
    CONFIGURE = "configure"
    VERIFY = "verify"


class Deployment(Base):
    __tablename__ = "deployments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    status = Column(
        Enum(DeploymentStatus, values_callable=lambda x: [e.value for e in x]),
        nullable=False,
        default=DeploymentStatus.PENDING,
    )
    current_step = Column(
        Enum(DeploymentStep, values_callable=lambda x: [e.value for e in x]),
        nullable=True,
    )
    progress = Column(Integer, nullable=False, default=0)
    platform = Column(String(20), nullable=False)
    config = Column(JSONB)
    git_url = Column(String(500))
    branch = Column(String(200), default="main")
    image_tag = Column(String(500))
    deploy_url = Column(String(500))
    commit_hash = Column(String(40))
    commit_message = Column(Text)
    error_message = Column(Text)
    error_details = Column(JSONB)
    can_resume = Column(Integer, default=0)
    resume_data = Column(JSONB)
    duration = Column(Integer)
    started_at = Column(DateTime)
    completed_at = Column(DateTime)
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    checkpoints = relationship("DeploymentCheckpoint", back_populates="deployment", cascade="all, delete-orphan")
    logs = relationship("DeploymentLog", back_populates="deployment", cascade="all, delete-orphan")


class DeploymentCheckpoint(Base):
    __tablename__ = "deployment_checkpoints"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(UUID(as_uuid=True), ForeignKey("deployments.id"), nullable=False)
    step = Column(String(50), nullable=False)
    step_index = Column(Integer, nullable=False)
    state_data = Column(JSONB)
    resources_created = Column(JSONB)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    deployment = relationship("Deployment", back_populates="checkpoints")


class DeploymentLog(Base):
    __tablename__ = "deployment_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    deployment_id = Column(UUID(as_uuid=True), ForeignKey("deployments.id"), nullable=False)
    level = Column(String(20), nullable=False, default="info")
    message = Column(Text, nullable=False)
    details = Column(JSONB)
    step = Column(String(50))
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    deployment = relationship("Deployment", back_populates="logs")

File: services/api/src/models/deployment.rs
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, EnumIter, DeriveActiveEnum, Serialize, Deserialize)]
#[sea_orm(rs_type = "String", db_type = "Enum", enum_name = "deploymentstatus")]
#[serde(rename_all = "snake_case")]
pub enum DeploymentStatus {
    #[sea_orm(string_value = "pending")]
    Pending,
    #[sea_orm(string_value = "running")]
    Running,
    #[sea_orm(string_value = "paused")]
    Paused,
    #[sea_orm(string_value = "cancelled")]
    Cancelled,
    #[sea_orm(string_value = "success")]
    Success,
    #[sea_orm(string_value = "failed")]
    Failed,
    #[sea_orm(string_value = "rolling_back")]
    RollingBack,
    #[sea_orm(string_value = "rolled_back")]
    RolledBack,
    #[sea_orm(string_value = "waiting_review")]
    WaitingReview,
}

#[derive(Debug, Clone, PartialEq, Eq, EnumIter, DeriveActiveEnum, Serialize, Deserialize)]
#[sea_orm(rs_type = "String", db_type = "Enum", enum_name = "deploymentstep")]
#[serde(rename_all = "snake_case")]
pub enum DeploymentStep {
    #[sea_orm(string_value = "clone")]
    Clone,
    #[sea_orm(string_value = "generate_review")]
    GenerateReview,
    #[sea_orm(string_value = "build")]
    Build,
    #[sea_orm(string_value = "env_review")]
    EnvReview,
    #[sea_orm(string_value = "push")]
    Push,
    #[sea_orm(string_value = "deploy")]
    Deploy,
    #[sea_orm(string_value = "configure")]
    Configure,
    #[sea_orm(string_value = "verify")]
    Verify,
}

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "deployments")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Uuid,
    pub project_id: Uuid,
    pub user_id: Option<Uuid>,
    pub status: DeploymentStatus,
    pub current_step: Option<DeploymentStep>,
    pub progress: i32,
    pub platform: String,
    pub config: Option<Json>,
    pub git_url: Option<String>,
    pub branch: Option<String>,
    pub image_tag: Option<String>,
    pub deploy_url: Option<String>,
    pub commit_hash: Option<String>,
    pub commit_message: Option<String>,
    pub error_message: Option<String>,
    pub error_details: Option<Json>,
    pub can_resume: Option<i32>,
    pub resume_data: Option<Json>,
    pub duration: Option<i32>,
    pub started_at: Option<chrono::NaiveDateTime>,
    pub completed_at: Option<chrono::NaiveDateTime>,
    pub updated_at: Option<chrono::NaiveDateTime>,
    pub created_at: Option<chrono::NaiveDateTime>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::project::Entity",
        from = "Column::ProjectId",
        to = "super::project::Column::Id"
    )]
    Project,
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::UserId",
        to = "super::user::Column::Id"
    )]
    User,
}

impl Related<super::project::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Project.def()
    }
}

impl Related<super::user::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
